rfc/rfc-43/rfc-43.md

```diff
@@ -0,0 +1,222 @@
+   2. With the increase in the number of HUDI jobs, there is no unified service to manage compaction/clustering jobs (monitor, retry, history, etc...), which will make maintenance costs increase.
+
+With this effort, we want to provide an independent compaction/clustering Service, it will have these abilities:
+
+- Provides a pluggable execution interface that can adapt to multiple execution engines, such as Spark and Flink.
+
+- With the ability to failover, need to be persisted compaction/clustering message.
+
+- Perfect metrics and reuse HoodieMetric expose to the outside.
+
+- Provide automatic failure retry for compaction/clustering job.
+
+## Implementation
+
+![](service.png)
+
+### Client
+
+Register to Service with CompactionConfig/ClusteringConfig when the job starts. When generating a compaction/clustering plan, report to the compaction/clustering Service address specified in HoodieConfig and request the Service to stop the corresponding compaction/clustering job when the Service is not required or rollback compaction/clustering instant.
+
+### Request Handler
+
+Receive the client's request and save the plan in the meta table.
+
+### Meta Table
+
+Meta table has been implemented by using an internal HUDI MOR Table to store the required metadata for the compaction/clustering message. This table will be internal to a dataset and will not be exposed directly to the user to write/modify.
+
+The Meta Table has the following features:
+
+1. Is a HUDI MOR Table
+
+2. Defines its own custom payload HoodieServicePayload
+
+3. Use basePath + instant as recordKey to guarantee unique
+
+4. Follows all the conventions of a HUDI Table
+
+### Scheduler
+
+Periodically scan the Meta Table and submit compaction/clustering job according to user-specified rules, and need to plug-in the execution engine.
+
+### Schema
+
+Being an actual Hoodie Table, the Meta Table needs to have a schema to store the compaction/clustering message. 
+
+```
+{
+    "namespace":"org.apache.hudi.avro.model",
+    "type":"record",
+    "name":"HoodieServiceRecord",
+    "doc":"Compaction/Clustering message saved within the Meta Table",
+    "fields":[
+        {
+            "name":"table_name",
+            "type":"string"
+        },
+        {
+            "name":"base_path",
+            "type":"string"
+        },
+        {
+            "name":"instant",
+            "doc":"compaction/clustering instant for the hoodie table",
+            "type":"string"
+        }, {
+            "name":"action",
+            "doc":"compact or cluster for this instant",
+            "type":"string"
+        },{
+            "name":"status",
+            "doc":"Status of the compaction/clustering job, 0：Scheduled 1：Running 2：Retrying 3:  Completed",
+            "type":"int"
+        } ,{
+            "name":"isDeleted",
+            "doc":"True if this plan has been deleted",
+            "type":"int"
+        }
+    ]
+}
+```
+
+## Interface design
+
+### Register
```

Comment:
What is the different between register and schedule?

I feel we can merge the two to just register:
 When you register a job, it will be added to the list of jobs and scheduled as necessary. Some jobs may be periodic (e.g. clean and compaction) and some jobs one-shot (e.g. clustering).

Reply:
from my understanding, actual scheduling (strategy of when to compact and which file groups to compact) is not delegated to this service. That still lies w/ original hudi table. Once compaction plan is finalized, we delegate it to this service. 

So, not sure if we need to overload this service w/ actual scheduling as well. But I do see there is scope of improvement and add the functionality to the service. may be in first cut, we can leave it to original hudi table.

Reply:
Registration means that this table is scheduled by the service. This interface is to reserve the scheme of subsequent calls through the service, which will be a follow up.In the current first version we can leave the scheduling to the original HUDI table.

@prashantwason @nsivabalan What do you think?

Reply:
> Registration means that this table is scheduled by the service. 

Each table that configured to use TMS is already deemed as registered and for subsequent table service scheduling, it should happen automatically (via some other url config?). I don't see it's a must to provide a special API for registration.